### PR TITLE
Initial integration of builds_status

### PR DESCRIPTION
**Pull request by kabeer27**
Changes
-added logs bucket to build_project/ request_build
-added new datastore entity for storing build history of projects

Added update_builds_status function:
-Changed functionality of get_last_build by using cloud datastore
-refactored existing code to make some of it reusable.
-fixed all linting errors for builds_status.py
-Badge images are now downloaded from github and stored in /tmp folder available in cloud function's runtime.

@oliverchang 
@Dor1s 
@inferno-chromium

### CODE REVIEW

File: infra/build/functions/datastore_entities.py [hunk 1]
```diff
@@ -30,3 +30,10 @@ class Project(ndb.Model):
 class GitAuth(ndb.Model):
   """Represents Github access token entity."""
   access_token = ndb.StringProperty()
+
+
+# pylint: disable=too-few-public-methods
+class BuildsHistory(ndb.Model):
+	tag = ndb.StringProperty()
```
Comment (oliverchang) [inline infra/build/functions/datastore_entities.py:37]: fix indentation to be consistent. Also rename this to build_tag_suffix.

File: infra/build/functions/request_build.py [hunk 2]
```diff
@@ -23,11 +23,26 @@
...
 
 
+def update_build_history(project_name, build_id, tag):
+  """Update build history of project."""
+  build_history = BuildsHistory.query(BuildsHistory.project==project_name, BuildsHistory.tag==tag)
+  project = build_history.get()
+  if project is None:
+    project = BuildsHistory(tag=tag, project=project_name, build_ids=[])
+
+  if len(project.build_ids) > 63:
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:39]: Use 64 and put it into a constant. MAX_BUILD_HISTORY_LENGTH.

File: infra/build/functions/update_build_status.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+# pylint: disable=no-member
+def update_status(event, context):
+  """Entry point for cloud function to update build statuses and badges."""
+  with ndb.Client().context():
+    builds = BuildsHistory.query()
+
+  projects = [build.project for build in builds if build.tag=='-fuzzing']
+
+  #TODO; Cleanup after infrastructure migration
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:65]: clean up what? be specific. Also use ":" not ";"

File: infra/build/functions/update_build_status.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      build_project.FUZZING_BUILD_TAG,
+      status_filename='status.json')
+  builds_status.update_build_status(
+      builds
+      projects,
+      build_and_run_coverage.COVERAGE_BUILD_TAG,
+      status_filename='status-coverage.json')
+
+  for project in projects:
+    last_co
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:80]: remove?

File: infra/gcb/builds_status.py [hunk 5]
```diff
@@ -236,10 +248,10 @@ def main():
       status_filename='status-coverage.json')
 
   update_build_badges(
-      builds,
-      projects,
-      build_tag=build_project.FUZZING_BUILD_TAG,
-      coverage_tag=build_and_run_coverage.COVERAGE_BUILD_TAG)
+    builds,
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:251]: I don't think this new indentation is correct.

File: infra/build/functions/update_build_status.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+# pylint: disable=no-member
+def update_status(event, context):
+  """Entry point for cloud function to update build statuses and badges."""
+  with ndb.Client().context():
+    builds = BuildsHistory.query()
+
+  projects = [build.project for build in builds if build.tag=='-fuzzing']
+
+  #TODO; Cleanup after infrastructure migration
+  builds_status.find_last_build = get_last_build
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:66]: Don't do this. If you need to change some behaviour, pass it as an argument rather than monkey patching.

File: infra/build/functions/update_build_status.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import build_and_run_coverage
+import build_project
+import builds_status
+from datastore_entities import BuildsHistory
+
+
+# TODO: update function in builds_status.py after migration.
+def get_last_build(builds, project, build_tag_suffix):
+  """Returns build object for the last finished build of project."""
+  with ndb.Client().context():
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:34]: Use a top level ndb context in the entry point of your function.

Does this even work as-is when you have .get() on line 38?

File: infra/build/functions/update_build_status.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    build_history = builds.filter(BuildsHistory.project == project)
+    build_history_with_tag = build_history.filter(BuildsHistory.tag == tag)
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:36]: where is "tag" defined? Also just put both filters in the one call:

```python
build_history = BuildsHistory.query(BuildsHistory.project == project, BuildsHistory.tag == tag)
```

File: infra/build/functions/request_build.py [hunk 9]
```diff
@@ -23,11 +23,26 @@
...
 
 BASE_PROJECT = 'oss-fuzz-base'
 
 
+def update_build_history(project_name, build_id, tag):
+  """Update build history of project."""
+  build_history = BuildsHistory.query(BuildsHistory.project==project_name, BuildsHistory.tag==tag)
+  project = build_history.get()
+  if project is None:
+    project = BuildsHistory(tag=tag, project=project_name, build_ids=[])
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:37]: Set a key here based on project_name + tag to avoid duplicates from races.

File: infra/build/functions/request_build.py [hunk 10]
```diff
@@ -23,11 +23,26 @@
...
+from datastore_entities import BuildsHistory
 from datastore_entities import Project
 
 BASE_PROJECT = 'oss-fuzz-base'
 
 
+def update_build_history(project_name, build_id, tag):
+  """Update build history of project."""
+  build_history = BuildsHistory.query(BuildsHistory.project==project_name, BuildsHistory.tag==tag)
+  project = build_history.get()
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:35]: "project" is not a the right variable name here. Use "build_history", and rename the query to "build_history_query".

File: infra/build/functions/update_build_status.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+  credentials, image_project = google.auth.default()
+  cloudbuild = build('cloudbuild', 'v1', credentials=credentials)
+
+  for build_id in reversed(project_build_history.build_ids):
+    build = cloudbuild.projects().builds().get(projectId=image_project, id=build_id).execute()
+    if build['status'] == 'WORKING':
+      continue
+
+    if not builds_status.upload_log(build_id):
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:50]: We shouldn't repeatedly upload a log if it's already uploaded. You should add a check in upload_log for that.

File: infra/build/functions/update_build_status.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    return build
+
+  return None
+
+
+# pylint: disable=no-member
+def update_status(event, context):
+  """Entry point for cloud function to update build statuses and badges."""
+  with ndb.Client().context():
+    builds = BuildsHistory.query()
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:61]: This makes the code a bit harder to follow since they're passed off to other functions. Just initialise this where they are used (i.e. update_build_status).

File: infra/gcb/builds_status.py [hunk 13]
```diff
@@ -183,37 +223,32 @@ def update_build_status(builds, projects, build_tag_suffix, status_filename):
...
-    last_coverage_build = find_last_build(builds, project, coverage_tag)
-    if not last_build or not last_coverage_build:
-      continue
-
-    badge = 'building'
-    if not is_build_successful(last_coverage_build):
-      badge = 'coverage_failing'
-    if not is_build_successful(last_build):
-      badge = 'failing'
+def update_build_badges(project, last_build, last_coverage_build):
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:226]: Update this so that it accepts `last_build_successful` and `last_coverage_build_successful` instead of expecting an entire build dict.

File: infra/build/functions/update_build_status.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+      coverage_build_history = coverage_build_history_query.get()
+      if not coverage_build_history:
+        continue
+
+      last_coverage_build = get_last_build(coverage_build_history.build_ids)
+      if not last_coverage_build:
+        continue
+
+      builds_status.update_build_badges(project.name, last_build,
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:114]: Rather than do all the stuff in lines 91-115 again, we can improve things over the initial implementation by getting update_build_status to return the information we need. 

e.g. it can return a dict of project name -> build success, and can be used to make the builds_status.update_build_badges call.

File: infra/build/functions/update_build_status.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+  cloudbuild = build('cloudbuild', 'v1', credentials=credentials)
+
+  for build_id in reversed(build_ids):
+    project_build = cloudbuild.projects().builds().get(projectId=image_project,
+                                                       id=build_id).execute()
+    if project_build['status'] == 'WORKING':
+      continue
+
+    if not builds_status.upload_log(build_id):
+      continue
```
[the remark this answers is not in the document]
Comment (kabeer27) [reply infra/build/functions/update_build_status.py:43]: the function returns False in case the build log with the specified build id does not exist in our gcb-logs-bucket, so if the log file is missing it is better not to update it since it won't be copied over to our bucket in clusterfuzz

[the remark this answers is not in the document]
Comment (oliverchang) [reply infra/build/functions/update_build_status.py:43]: Raise an exception in that case. That would be a critical error in our infrastructure.

[the remark this answers is not in the document]
Comment (kabeer27) [reply infra/build/functions/update_build_status.py:43]: Alright will do

File: infra/build/functions/request_build.py [hunk 16]
```diff
@@ -23,9 +23,29 @@
...
 BASE_PROJECT = 'oss-fuzz-base'
+MAX_BUILD_HISTORY_LENGTH = 64
+
+
+def update_build_history(project_name, build_id, build_tag_suffix):
+  """Update build history of project."""
+  project_key = ndb.Key(BuildsHistory, project_name + build_tag_suffix)
+  project = project_key.get()
+
+  if project is None:
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:38]: nit: just `if not project`

File: infra/build/functions/update_build_status.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+  for build_id in reversed(build_ids):
+    project_build = cloudbuild.projects().builds().get(projectId=image_project,
+                                                       id=build_id).execute()
+    if project_build['status'] == 'WORKING':
+      continue
+
+    if not builds_status.upload_log(build_id):
+      log_name = 'log-{0}'.format(build_id)
+      raise MissingBuildLogError(
+          'Critical Error Missing build log file {0} in cloud storage'.format(
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:49]: Just "Missing build log file {0}".

File: infra/build/functions/request_build.py (hunk 16 above)
Comment (oliverchang) [reply infra/build/functions/request_build.py:38]: This didn't seem addressed.

File: infra/build/functions/request_build.py [hunk 18]
```diff
@@ -23,17 +23,37 @@
...
+  project.put()
 
 
 def get_project_data(project_name):
   """Retrieve project metadata from datastore."""
   with ndb.Client().context():
     query = Project.query(Project.name == project_name)
     project = query.get()
-    if project is None:
+    if not Project:
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:56]: Project shouldn't be capitalised. Otherwise this check isn't working as expected. Please add a test here as well so this is caught.

File: infra/build/functions/update_build_status.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+
+class MissingBuildLogError(Exception):
+  """Missing build log file in cloud storage."""
+
+
+def download_badge_images():
+  """Download badge images from github into a /tmp."""
+  if not os.path.exists('/tmp/badge_images'):
+    os.mkdir('/tmp/badge_images')
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:44]: os.makedirs('/tmp/badge_images', exist_ok=True)

File: infra/build/functions/update_build_status.py [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+  builds_status.upload_status(successes, failures, status_filename)
+  return statuses
+
+
+# pylint: disable=no-member
+def update_status(event, context):
+  """Entry point for cloud function to update build statuses and badges."""
+  del event, context  #unused
+  builds_status.SCRIPT_DIR = '/tmp'
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:115]: don't do this. instead, as before with another case, refactor/pass this as an argument.

File: infra/build/functions/update_build_status.py [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+from google.cloud import ndb
+
+import build_and_run_coverage
+import build_project
+import builds_status
+from datastore_entities import BuildsHistory
+from datastore_entities import Project
+
+# pylint: disable=line-too-long
+BADGE_IMAGES_BASE_URL = "https://raw.githubusercontent.com/google/oss-fuzz/master/infra/gcb/badge_images/"
```
Comment (oliverchang) [inline infra/build/functions/update_build_status.py:33]: single quotes. also, to avoid the line-too-long lint error, break up the string like e.g.

```python
... = (
   'abc'
   'def'
)
```

File: infra/gcb/builds_status.py [hunk 22]
```diff
@@ -66,16 +89,40 @@ def upload_status(successes, failures, status_filename):
...
+
+  if not log.exists():
+    print('Failed to find build log {0}'.format(log_name), file=sys.stderr)
+    return False
+
+  if dest_log.exists():
+    return True
+
+  with tempfile.NamedTemporaryFile() as file:
+    log.download_to_filename(file.name)
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:112]: Let's fix this to use bucket.copy_blob so we don't have have to download the file. See https://googleapis.dev/python/storage/latest/buckets.html

File: infra/gcb/builds_status.py [hunk 23]
```diff
@@ -59,23 +81,45 @@ def upload_status(successes, failures, status_filename):
...
+
+  if not log.exists():
+    print('Failed to find build log {0}'.format(log_name), file=sys.stderr)
+    return False
+
+  if dest_log.exists():
+    return True
+
+  gcb_bucket.copy_blob(log, status_bucket)
+
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:111]: remove unnecessary newline

File: infra/gcb/builds_status.py [hunk 24]
```diff
@@ -24,28 +42,32 @@
 BUILDS_PAGE_SIZE = 256
 BADGE_IMAGE_TYPES = {'svg': 'image/svg+xml', 'png': 'image/png'}
 
-_client = None
+_CLIENT = None
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:45]: this shouldn't be all caps, as it's not a constant. Did lint complain here? If so, let's add a ignore.

File: infra/gcb/builds_status.py [hunk 25]
```diff
@@ -1,10 +1,28 @@
+# Copyright 2020 Google Inc.
```
Comment (oliverchang) [inline infra/gcb/builds_status.py:1]: let's revert most of the changes here, as you've reimplemented most of the functionality.

File: infra/gcb/builds_status.py (hunk 24 above)
Comment (kabeer27) [reply infra/gcb/builds_status.py:45]: Yeah lint complained, adding an ignore

File: infra/build/functions/requirements.txt [hunk 26]
```diff
@@ -20,6 +20,9 @@ grpcio==1.29.0
 google-auth==1.18.0
 google-cloud-ndb==1.3.0
 google-cloud-scheduler==1.3.0
+google-cloud-storage==1.29.0
 google-api-core==1.21.0
 google-api-python-client==1.9.3
 oauth2client==4.1.3
+python-dateutil==2.8.1 
+wget==3.2
```
Comment (oliverchang) [inline infra/build/functions/requirements.txt:28]: why is this needed?
